pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+	rbs, err := grb.roleBindingLister.List("", labels.NewSelector().Add(*r))
+	if err != nil {
+		returnError = multierror.Append(returnError,
+			errors.Wrapf(err, "couldn't list roleBindings with label %s : %s", grbOwnerLabel, globalRoleBinding.Name))
+	} else {
+		err = grb.purgeInvalidRoleBindings(rbs, gr.NamespacedRules)
+		if err != nil {
+			returnError = multierror.Append(returnError, err)
+		}
+	}
+
+	for ns := range gr.NamespacedRules {
+		_, err := grb.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+
+		// Check if the role we're binding to exists
+		roleName := gr.Name + "-" + ns
+		_, err = grb.roleLister.Get(ns, roleName)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get role %s", roleName))
+			continue
+		}
+
+		subject := rbac.GetGRBSubject(globalRoleBinding)
+		rbName := globalRoleBinding.Name + "-" + ns
+		roleBinding, err := grb.roleBindingLister.Get(ns, rbName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+
+			newRoleBinding := &v1.RoleBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      rbName,
+					Namespace: ns,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: globalRoleBinding.APIVersion,
+							Kind:       globalRoleBinding.Kind,
+							UID:        globalRoleBinding.UID,
+							Name:       globalRoleBinding.Name,
+						},
+					},
+					Labels: map[string]string{
+						grbOwnerLabel: globalRoleBinding.Name,
+					},
+				},
+				RoleRef: v1.RoleRef{
+					Name: roleName,
+					Kind: "Role",
+				},
+				Subjects: []v1.Subject{subject},
+			}
+			_, err := grb.roleBindings.Create(newRoleBinding)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+		} else {
+			// Check that the RoleBinding refers to the right Role
+			roleRef := v1.RoleRef{
+				Name: roleName,
+				Kind: "Role",
+			}
+			if !reflect.DeepEqual(roleRef, roleBinding.RoleRef) || roleBinding.Labels[grbOwnerLabel] != globalRoleBinding.Name {
+				newRoleBinding := roleBinding.DeepCopy()
+				newRoleBinding.RoleRef = roleRef
+				if newRoleBinding.Labels != nil {
+					newRoleBinding.Labels[grbOwnerLabel] = globalRoleBinding.Name
+				} else {
+					newRoleBinding.Labels = map[string]string{grbOwnerLabel: globalRoleBinding.Name}
+				}
+				_, err = grb.roleBindings.Update(newRoleBinding)
+				if err != nil {
+					returnError = multierror.Append(returnError, err)
+					continue
+				}
+			}
+		}
+	}
+
+	return returnError
+}
+
+// remove any roleBindings that aren't in the namespaces listed in the associated GlobalRole.namespacedRules
+func (grb *globalRoleBindingLifecycle) purgeInvalidRoleBindings(rbs []*v1.RoleBinding, namespacedRules map[string][]v1.PolicyRule) error {
```

Comment:
Nit: If you could name this in a way that distinguishes it from the other purge method (which is only for the inheritedClusterRole logic), that would be preferable.